Mathlib/RingTheory/HahnSeries.lean

```diff
@@ -618,22 +618,79 @@ theorem order_one [MulZeroOneClass R] : order (1 : HahnSeries Γ R) = 0 := by
   · exact order_single one_ne_zero
 #align hahn_series.order_one HahnSeries.order_one
 
-instance [NonUnitalNonAssocSemiring R] : Mul (HahnSeries Γ R) where
-  mul x y :=
-    { coeff := fun a =>
-        ∑ ij in addAntidiagonal x.isPWO_support y.isPWO_support a, x.coeff ij.fst * y.coeff ij.snd
-      isPWO_support' :=
+variable {V : Type*} [AddCommMonoid V]
+
+/-- We introduce a type alias for HahnSeries in order to work with scalar multiplication by series.
+If we wrote a `SMul (HahnSeries Γ R) (HahnSeries Γ V)` instance, then when `V = HahnSeries Γ R`, we
+would have two different actions of `HahnSeries Γ R` on `HahnSeries Γ V`.  See
+`Mathlib.Data.Polynomial.Module` for more discussion on this problem. -/
+@[nolint unusedArguments]
+def _root_.HahnModule (Γ R V : Type*) [PartialOrder Γ] [Zero V] [SMul R V] :=
+  HahnSeries Γ V
+
+@[ext]
+theorem _root_.HahnModule.ext {Γ R V : Type*} [PartialOrder Γ] [Zero V] [SMul R V]
+    (x y : HahnModule Γ R V) (coeff : x.coeff = y.coeff) : x = y := by simp_all only [coeff_inj]
+
+instance instZeroHahnModule [SMul R V] : Zero (HahnModule Γ R V) :=
+  ⟨{  coeff := 0
+      isPWO_support' := by simp }⟩
+
+instance instAddHahnModule [SMul R V] : Add (HahnModule Γ R V) where
+  add x y :=
+    { coeff := x.coeff + y.coeff
+      isPWO_support' := (x.isPWO_support.union y.isPWO_support).mono (Function.support_add _ _) }
+
+instance instAddCommMonoidHahnMonoid [SMul R V] : AddCommMonoid (HahnModule Γ R V) where
+  zero := 0
+  add := (· + ·)
+  add_assoc x y z := by
+    ext
+    apply add_assoc
+  zero_add x := by
+    ext
+    apply zero_add
+  add_zero x := by
+    ext
+    apply add_zero
+  add_comm := fun x y => by
+    ext
+    apply add_comm
+
+@[simp]
+theorem add_coeff'' [SMul R V] {x y : HahnModule Γ R V} {a : Γ} :
+    (x + y).coeff a = x.coeff a + y.coeff a := rfl
+
+instance instSMulHahn [Zero R] [SMul R V] : SMul (HahnSeries Γ R)
+    (HahnModule Γ R V) where
```

Comment:
I think it would be good to have all of this in a `namespace HahnModule`. Could you please refactor the file a bit to make this possible?

Reply:
Sounds good